katran/lib/bpf/balancer_kern.c

```diff
@@ -628,6 +628,11 @@ static inline int process_packet(void *data, __u64 off, void *data_end,
   }
   data_stats->v1 += 1;
   data_stats->v2 += pkt_bytes;
+#ifdef LOCAL_DELIVERY_OPTIMIZATION
+  if ((vip_info->flags & F_LOCAL_VIP) && (dst->flags & F_LOCAL_REAL)) {
+    return XDP_PASS;
+  }
+#endif
```

Comment:
some additional thoughs. lines 607-615 should be moved after this new section. there is no point to do encapsulation if we are going to do XDP_PASS (in this case linux kernel wont need to do decapsulation)

Reply:
If we move these lines after line 635, the stats will be updated in case of encapsulation return false (XDP_DROP). Is this OK?

Reply:
the stats should be ok since they are counting packets received per vip and per backend

Reply:
well they are also coutning "per real". but as i mentioned bellow it is ok